### PR TITLE
OF.TV extractors

### PR DESCRIPTION
**Pull request by DoubleCouponDay**
### Description of your *pull request* and other information

</details>

This feature adds support for the video streaming website [https://OF.TV](https://OF.TV). Two extractors have been created for single videos and entire playlists. While the generic extractor can handle this website, it extracts irrelevant videos.

Fixes: #5017

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [X] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/oftv.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            'view_count': 0
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        extraction = ZypeIE.extract_from_webpage(self._downloader, url, webpage)
+        output = list(extraction)[0]
+        return output
```
Comment (pukkandan) [inline yt_dlp/extractor/oftv.py:31]:
```suggestion
        return next(ZypeIE.extract_from_webpage(self._downloader, url, webpage))
```

File: yt_dlp/extractor/oftv.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    _TESTS = [{
+        'url': 'https://of.tv/video/627d7d95b353db0001dadd1a',
+        'md5': 'cb9cd5db3bb9ee0d32bfd7e373d6ef0a',
+        'info_dict': {
+            'id': '627d7d95b353db0001dadd1a',
+            'ext': 'mp4',
+            'title': 'E1: Jacky vs Eric',
+            'thumbnail': r're:^https?://.*\.jpg',
+            'average_rating': 0,
+            'description': 'Singer Jacky Romero and actor Eric Guilmette must step out of their comfort zones in order to take on the ultimate comfort food: the Monte Cristo sandwich. With only 30 minutes to deliver the perfect blend of sweet and savory,\xa0Jacky and Eric need all the help they can get to impress Chef JoJo and move on to the next round.\xa0\r\n\r\nSubscribe to the cast on OF:\r\nof.com/jackyromero \r\nof.com/eric_g \r\nof.com/grandmasterchefjojo',
```
Comment (pukkandan) [inline yt_dlp/extractor/oftv.py:17]: use md5

File: yt_dlp/extractor/oftv.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        'info_dict': {
+            'id': 'this-is-fire'
+        }
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        playlists_match = self._search_regex(r'var\s*remaining_videos\s*=\s*(\[.+?\])\s*;', webpage, 'oftv playlists')
+        remaining_videos = self._parse_json(playlists_match, playlist_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/oftv.py:48]: merge with master and then do:

```suggestion
        remaining_videos = self._search_json(
            r'var\s*remaining_videos\s*=', webpage, 'oftv playlists', playlist_id, contains_regex=r'\[.+\]')
```

File: yt_dlp/extractor/oftv.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        filtered = traverse_obj(remaining_videos, (..., "discovery_url", ))
+        return self.playlist_from_matches(filtered, playlist_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/oftv.py:50]:
```suggestion
        return self.playlist_from_matches(
            traverse_obj(remaining_videos, (..., 'discovery_url')), playlist_id)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by bashonly** [comment]
besides playlist extraction, wasn't the other reason for an `of.tv` extractor so that it could extract more metadata (specifically creator/uploader) since the generic extractor was already extracting the zype video info?

**Comment by pukkandan** [comment]
Yes. As it is rn, this will only partially satisfy the linked issue

**Comment by DoubleCouponDay** [comment]
Sweet I'll fix all those problems tonight after work.

File: yt_dlp/extractor/oftv.py (hunk 1 above)
Comment (DoubleCouponDay) [reply yt_dlp/extractor/oftv.py]: completed in 4d7c53bdee6b80086418ff29de9ea5c4f8b88ad0

File: yt_dlp/extractor/oftv.py (hunk 3 above)
Comment (DoubleCouponDay) [reply yt_dlp/extractor/oftv.py:46]: completed in 7ad7e73705906980e63289baaa683da73d1091c0

File: yt_dlp/extractor/oftv.py (hunk 2 above)
Comment (DoubleCouponDay) [reply yt_dlp/extractor/oftv.py:17]: was this commit what you were looking for? e7d2ee9063ca06b9b1ea054dc5e7c2042c82c72e

File: yt_dlp/extractor/oftv.py (hunk 4 above)
Comment (DoubleCouponDay) [reply yt_dlp/extractor/oftv.py:48]: completed in 5745f525028eeb45c0b52330f1fff8c9a800ffa1

File: yt_dlp/extractor/oftv.py (hunk 2 above)
Comment (bashonly) [reply yt_dlp/extractor/oftv.py:17]: for test fields where the string value is long/unwieldy you can use an md5 hash in place of the string
```py
            'description': 'md5:TODO',
```
run the test with this and the result will give you a hash to replace `TODO` with

Comment (DoubleCouponDay) [reply yt_dlp/extractor/oftv.py:17]: completed in 0e46566531b75a8650f955075b65fb72b0a15f25

**Pull request closed without merging**